art/defences/detector/poison/clustering_centroid_analysis.py

```diff
@@ -0,0 +1,555 @@
+    Extract features from the model using the feature representation sub model.
+
+    :param x_train: numpy array d-dimensional features for n data entries. Features are extracted from here
+    :param feature_representation_model: DNN submodel from input up to feature abstraction
+    :return: features. numpy array of features
+    """
+    # Convert data to TensorFlow tensors if needed
+    data = x_train
+    if not isinstance(x_train, tf.Tensor):
+        data = tf.convert_to_tensor(x_train, dtype=tf.float32)
+
+    # Process in batches to avoid memory issues
+    batch_size = 256
+    num_batches = int(np.ceil(len(data) / batch_size))
+    features: list[tf.Tensor] = []
+
+    for i in range(num_batches):
+        start_idx = i * batch_size
+        end_idx = min((i + 1) * batch_size, len(data))
+        batch = data[start_idx:end_idx]
+        batch_features = _calculate_features(feature_representation_model, batch)
+        features.append(batch_features)
+
+    # Concatenate all batches
+    final_features_tensor: tf.Tensor
+    if len(features) > 1:
+        final_features_tensor = tf.concat(features, axis=0)
+    else:
+        final_features_tensor = features[0]
+
+    return final_features_tensor.numpy()
+
+
+def _cluster_classes(
+    y_train: np.ndarray,
+    unique_classes: set[int],
+    features: np.ndarray,
+    clusterer: ClusterMixin,
+) -> tuple[np.ndarray, dict]:
+    """
+    Clusters all the classes in the given dataset into uniquely identifiable clusters.
+
+    :param y_train: numpy array of labels for n data entries
+    :param unique_classes: set of unique classes
+    :param features: feature representations' array of n rows
+    :param clusterer: clustering algorithm used
+    :return: (class_cluster_labels, cluster_class_mapping)
+    """
+    # represents the number of clusters used up until now to differentiate clusters obtained in different
+    # clustering runs by classes
+    logging.info("Clustering classes...")
+    used_cluster_labels = 0
+    cluster_class_mapping = {}
+    class_cluster_labels = np.full(len(y_train), -1)
+
+    logging.debug("Unique classes are: %s", unique_classes)
+
+    for class_label in unique_classes:
+        cluster_labels, selected_indices = _class_clustering(
+            y_train, features, class_label, clusterer
+        )
+        # label values are adjusted to account for labels of previous clustering tasks
+        cluster_labels[cluster_labels != -1] += used_cluster_labels
+        used_cluster_labels += len(np.unique(cluster_labels[cluster_labels != -1]))
+
+        class_cluster_labels[selected_indices] = cluster_labels
+
+        # the class (label) corresponding to the cluster is saved for centroid deviation calculation
+        for label in np.unique(cluster_labels[cluster_labels != -1]):
+            cluster_class_mapping[label] = class_label
+
+    return class_cluster_labels, cluster_class_mapping
+
+
+class ClusteringCentroidAnalysis(PoisonFilteringDefence):
```

Comment:
Change name because it is a framework specific implementation.
```suggestion
class ClusteringCentroidAnalysisTensorFlowV2(PoisonFilteringDefence):
```